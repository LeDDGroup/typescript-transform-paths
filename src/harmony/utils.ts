export type DownSampleTsTypes<TypeMap extends [unknown, unknown][], Tuple extends [...unknown[]]> = {
  [i in keyof Tuple]: Tuple[i] extends unknown[]
    ? DownSampleTsTypes<TypeMap, Tuple[i]>
    : DownSampleTsType<TypeMap, Tuple[i]>;
} & {
  length: Tuple["length"];
};

type DownSampleTsType<TypeMap extends [unknown, unknown][], T> =
  T extends Exclude<TypeMap[number][0], undefined> ? Extract<TypeMap[number], [T, unknown]>[1] : T;
