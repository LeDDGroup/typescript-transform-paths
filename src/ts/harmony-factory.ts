import TS from 'typescript';

/* ****************************************************************************************************************** */
// region: Types
/* ****************************************************************************************************************** */

export interface HarmonyFactory extends TS.NodeFactory {}

// endregion

/* ****************************************************************************************************************** */
// region: Utilities
/* ****************************************************************************************************************** */

// Leave this for now, as it serves as an example for when factory diverges again

// function getTs3Prop(instance: typeof TsThree, prop: string | symbol) {
//   switch (prop) {
//     case "updateCallExpression":
//       return (...args: any) => instance.updateCall.apply(void 0, args);
//     case "updateImportClause":
//       return function (
//         node: TsThree.ImportClause,
//         isTypeOnly: boolean,
//         name: TsThree.Identifier | undefined,
//         namedBindings: TsThree.NamedImportBindings | undefined
//       ) {
//         return instance.updateImportClause(node, name, namedBindings);
//       };
//     case "createImportClause":
//       return function (
//         name: TsThree.Identifier | undefined,
//         isTypeOnly: boolean,
//         namedBindings: TsThree.NamedImportBindings | undefined
//       ) {
//         return instance.createImportClause(name, namedBindings);
//       };
//     case "updateExportDeclaration":
//       return function (
//         node: TsThree.ExportDeclaration,
//         decorators: readonly TsThree.Decorator[] | undefined,
//         modifiers: readonly TsThree.Modifier[] | undefined,
//         isTypeOnly: boolean,
//         exportClause: TsThree.NamedExports | undefined,
//         moduleSpecifier: TsThree.Expression | undefined
//       ) {
//         return instance.updateExportDeclaration(
//           node,
//           node.decorators,
//           node.modifiers,
//           exportClause,
//           moduleSpecifier,
//           // @ts-ignore - This was added in later versions of 3.x
//           node.isTypeOnly
//         );
//       };
//     case "createExportDeclaration":
//       return function (
//         decorators: readonly TsThree.Decorator[] | undefined,
//         modifiers: readonly TsThree.Modifier[] | undefined,
//         isTypeOnly: boolean,
//         exportClause: TsThree.NamedExports | undefined,
//         moduleSpecifier: TsThree.Expression | undefined
//       ) {
//         return instance.createExportDeclaration(
//           decorators,
//           modifiers,
//           exportClause,
//           moduleSpecifier,
//           // @ts-ignore - This was added in later versions of 3.x
//           isTypeOnly
//         );
//       };
//     default:
//       return (...args: any) => (<any>instance)[prop](...args);
//   }
// }

/**
 * Creates a factory capable of working with any TS version using modern parameters
 */
export function createHarmonyFactory(tsInstance: typeof TS, factory: TS.NodeFactory | undefined): HarmonyFactory {
  const [majorVer, minorVer] = tsInstance.versionMajorMinor.split('.');

  return new Proxy(factory ?? tsInstance, {
    get(target, prop) {
      // if (!factory) return getTs3Prop(cast<typeof TsThree>(tsInstance), prop);

      // For 4.0 - 4.4
      if (+majorVer < 4 || +minorVer < 5) return (<any>target)[prop];

      // TODO - Handle 4.5
      return (<any>target)[prop];
    },
  }) as HarmonyFactory;
}

// endregion
