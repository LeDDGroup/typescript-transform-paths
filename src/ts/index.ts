export * from './harmony-factory';
export * from './ts-helpers';
