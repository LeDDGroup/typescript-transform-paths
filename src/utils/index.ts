export * from "./general-utils";
export * from "./elide-import-export";
export * from "./resolve-path-update-node";
export * from "./ts-type-conversion";
