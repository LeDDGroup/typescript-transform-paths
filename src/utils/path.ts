// noinspection JSUnusedLocalSymbols,JSUnusedAssignment

import path from "path";
import { normalizePath, Pattern, removePrefix, removeSuffix, ResolvedModuleFull } from "typescript";

/* ****************************************************************************************************************** */
// region: Types
/* ****************************************************************************************************************** */

export interface OutputPathDetail {
  isImplicitExtension: boolean
  isExternalLibraryImport: boolean
  resolvedExt: string | undefined;
  resolvedPath: string;
  outputPath: string | undefined;
  suppliedExt: string | undefined;
  implicitPath?: string;
  /** Package name from package.json */
  packageName?: string;
  /** Package name as written (could differ due to npm aliasing) */
  suppliedPackageName?: string
  suppliedPackagePath?: string
  packagePath?: string;
  tsPathMatch?: string;
}

// endregion

/* ****************************************************************************************************************** */
// region: Helpers
/* ****************************************************************************************************************** */

const pkgRegex = /^((@[^/]+\/[^/@]+)|([^@/]+))(?:\/([^@]+?))?$/;

/** @internal */
function getPaths(supplied: string, resolved: string) {
  let endMatchPos = 0;
  for (let i = 0; i < supplied.length && i < resolved.length; i++) {
    if (supplied[i] !== resolved[i]) {
      endMatchPos = i;
      break;
    }
  }

  const outputPath = supplied.slice(0, endMatchPos);
  const implicitPath = resolved.slice(endMatchPos);

  return { outputPath, implicitPath };
}

function getPackagePaths(moduleName: string, suppliedExt: string, subModule: string) {
  const { 1: suppliedPackageName, 2: suppliedPackagePath } = pkgRegex.exec(moduleName)!;

  const packagePathNoExt = fixupPartialPath(suppliedPackagePath, suppliedExt);

  if (!subModule) return { outputPath: packagePathNoExt };

  const subModuleNoExt = fixupPartialPath(subModule, path.extname(subModule));

  return {
    ...getPaths(packagePathNoExt, subModuleNoExt),
    suppliedPackageName,
    suppliedPackagePath
  };
}

function getModulePaths(
  moduleName: string,
  suppliedExt: string,
  resolvedFileName: string,
  resolvedExt: string,
  tsPathMatch: string | Pattern | undefined
) {
  // In this case, the file ending is a fixed path, so no further information needs to be determined
  if (typeof tsPathMatch === "string" || tsPathMatch?.suffix) return { outputPath: void 0 };

  const resolvedFileNameNoExt = fixupPartialPath(resolvedFileName, resolvedExt);
  const modulePathNoExt = tsPathMatch?.prefix
    ? fixupPartialPath(removePrefix(moduleName, tsPathMatch?.prefix), suppliedExt)
    : fixupPath(moduleName, suppliedExt);

  return getPaths(modulePathNoExt, resolvedFileNameNoExt);
}

function getTsPathMatch(match: string | Pattern): string {
  return typeof match === "string" ? match : [match.prefix, match.suffix].join("*");
}

// endregion

/* ****************************************************************************************************************** */
// region: Utils
/* ****************************************************************************************************************** */

export function joinPaths(...paths: (string | undefined)[]): string {
  return normalizePath(path.join('', ...paths.filter(p => typeof p === 'string') as string[]));
}

/**
 * Remove leading or trailing slashes
 * @p path to fix
 * @extName extname to remove
 */
export function fixupPartialPath(p: string, extName?: string): string {
  p = p.replace(/^[/\\]*(.+?)[/\\]*$/g, "$1");
  if (extName && p.slice(-extName.length) === extName) p = p.slice(0, p.length - extName.length);
  return p;
}

/**
 * Remove trailing slashes
 * @p path to fix
 * @extName extname to remove
 */
export function fixupPath(p: string, extName?: string): string {
  p = p.replace(/^(.+?)[/\\]*$/g, "$1");
  if (extName && p.slice(-extName.length) === extName) p = p.slice(0, p.length - extName.length);
  return p;
}

/** @internal â€” Uses internal TS type */
export function getOutputPathDetail(
  moduleName: string,
  resolvedModule: ResolvedModuleFull,
  pathMatch: string | Pattern | undefined
): OutputPathDetail {
  moduleName = fixupPath(moduleName);
  let suppliedExtName = path.extname(moduleName);
  const suppliedBaseName = path.basename(moduleName);
  let suppliedBaseNameNoExt = path.basename(moduleName, suppliedExtName);

  const resolvedFileName = resolvedModule.originalPath ?? resolvedModule.resolvedFileName;
  const resolvedExtName = path.extname(resolvedFileName);
  const resolvedFileNameNoExt = removeSuffix(resolvedFileName, resolvedExtName);
  const resolvedBaseName = path.basename(resolvedFileName);
  const resolvedBaseNameNoExt = path.basename(resolvedFileName, suppliedExtName);

  const tsPathMatch = pathMatch && getTsPathMatch(pathMatch);

  const { isExternalLibraryImport, packageId } = resolvedModule;
  const packageName = packageId?.name;
  const packageFileName = packageId && fixupPartialPath(packageId.subModuleName);
  const packageExtName = packageFileName && path.extname(packageFileName);
  const packageFileNameNoExt = packageFileName && removeSuffix(packageFileName, resolvedExtName);
  const packageBaseName = packageFileName && path.basename(packageFileName);
  const packageBaseNameNoExt = packageFileName && path.basename(packageFileName, packageExtName);

  const effectiveResolvedFileName = packageFileNameNoExt || resolvedFileNameNoExt;
  const effectiveResolvedBaseName = packageBaseName || resolvedBaseName;
  const effectiveResolvedBaseNameNoExt = packageBaseNameNoExt || resolvedBaseNameNoExt;
  const effectiveResolvedExtName = packageExtName || resolvedExtName;

  // Detect and fix invalid extname due to implicit ext (ie. `file.accounting.ts` could decide `accounting` is the extension)
  if (suppliedExtName && effectiveResolvedBaseNameNoExt && suppliedBaseName && effectiveResolvedBaseNameNoExt === suppliedBaseName) {
    suppliedBaseNameNoExt = suppliedBaseName;
    suppliedExtName = "";
  }

  const isImplicitExtension = !suppliedExtName;

  const pathDetail = resolvedModule.isExternalLibraryImport
    ? getPackagePaths(moduleName, suppliedExtName, packageId!.subModuleName)
    : getModulePaths(moduleName, suppliedExtName, resolvedFileName, resolvedExtName, tsPathMatch);

  return {
    isImplicitExtension,
    isExternalLibraryImport: !!isExternalLibraryImport,
    resolvedExt: effectiveResolvedExtName,
    suppliedExt: suppliedExtName,
    resolvedPath: effectiveResolvedFileName,
    tsPathMatch,
    packageName,
    packagePath: packageFileName,
    ...pathDetail,
  };
}

// endregion
