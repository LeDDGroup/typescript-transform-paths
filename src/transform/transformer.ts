import path from 'path';
import type TS from 'typescript';
import {
  ManualTransformOptions,
  PathResolver,
  TransformerConfig,
  TransformerContext,
  TransformerExtras,
  TransformerOptions,
  VisitorContext,
} from '../types';
import { nodeVisitor } from './visitor';
import { checkTsSupport, createHarmonyFactory, createSyntheticEmitHost, getTsNodeRegistrationProperties } from '../ts';
import { Minimatch } from 'minimatch';
import { IndexChecker } from '../resolve/index-checker';

/* ****************************************************************************************************************** */
// region: Helpers
/* ****************************************************************************************************************** */

function getTsDetail(p: {
  program: TS.Program | undefined;
  extras: TransformerExtras | undefined;
  manualTransformOptions: ManualTransformOptions | undefined;
  transformationContext: TS.TransformationContext;
}) {
  let tsInstance: typeof TS;
  let compilerOptions: TS.CompilerOptions;
  let fileNames: readonly string[] | undefined;
  let isTranspileOnly = false;
  let isTsNode = false;

  const { program, extras, manualTransformOptions, transformationContext } = p;

  tsInstance = extras?.ts ?? require('typescript');
  checkTsSupport(tsInstance);

  compilerOptions = manualTransformOptions?.compilerOptions!;

  /* Handle ts-node */
  const tsNodeProps = getTsNodeRegistrationProperties(tsInstance);
  if (tsNodeProps) isTsNode = true;

  /* Fixup options based on transform type */
  if (program) {
    compilerOptions ??= Object.assign({}, program.getCompilerOptions(), tsNodeProps?.compilerOptions);
  } else if (manualTransformOptions) {
    fileNames = manualTransformOptions.fileNames;
  } else {
    if (!tsNodeProps)
      throw new Error(
        `Cannot transform without a Program, ts-node instance, or manual parameters supplied. ` +
          `Make sure you're using ts-patch or ts-node with transpileOnly.`
      );
    isTranspileOnly = true;
    compilerOptions = tsNodeProps.compilerOptions;
    fileNames = tsNodeProps.fileNames;
  }

  /* Get EmitHost */
  const getCanonicalFileName = tsInstance.createGetCanonicalFileName(tsInstance.sys.useCaseSensitiveFileNames);

  let emitHost = transformationContext.getEmitHost();
  if (!emitHost || (isTsNode && isTranspileOnly)) {
    if (!fileNames)
      throw new Error(
        `No EmitHost found and could not determine files to be processed. Please file an issue with a reproduction!`
      );
    emitHost = createSyntheticEmitHost(compilerOptions, tsInstance, getCanonicalFileName, fileNames as string[]);
  }

  const projectReferences = program?.getResolvedProjectReferences()?.map((r) => r?.sourceFile);

  return { tsInstance, compilerOptions, fileNames, isTranspileOnly, isTsNode, emitHost, projectReferences };
}

function loadResolver(p: string): PathResolver {
  const resolver = require(p).default;
  if (typeof resolver !== 'function')
    throw new Error(`Cannot load resolver! Resolver file must export a resolver function as default`);

  return resolver;
}

/** @internal - Used by UT */
export function getTransformerConfig(options?: TransformerOptions): TransformerConfig {
  return {
    ...options,
    outputIndexes: options?.outputIndexes || 'auto',
    outputExtensions: options?.outputExtensions || 'auto',
    usePaths: options?.usePaths ?? true,
  };
}

// endregion

/* ****************************************************************************************************************** */
// region: Transformer
/* ****************************************************************************************************************** */

export function transformer(
  program?: TS.Program,
  options?: TransformerOptions,
  extras?: TransformerExtras,
  /**
   * Supply if manually transforming with compiler API via 'transformNodes' / 'transformModule'
   */
  manualTransformOptions?: ManualTransformOptions
) {
  return (transformationContext: TS.TransformationContext) => {
    const config = getTransformerConfig(options);
    const { tsInstance, compilerOptions, isTranspileOnly, isTsNode, emitHost } = getTsDetail({
      program,
      extras,
      manualTransformOptions,
      transformationContext,
    });

    const rootDirs = compilerOptions.rootDirs?.filter(path.isAbsolute);
    const { configFile, paths } = compilerOptions;
    const resolver = config.resolver ? loadResolver(config.resolver) : void 0;

    // Older TS does not have this, so we type it accordingly to determine path parse method
    const tryParsePatterns: typeof TS.tryParsePatterns | undefined = tsInstance.tryParsePatterns;

    const tsTransformPathsContext: TransformerContext = {
      compilerOptions,
      config,
      elisionMap: new Map(),
      tsFactory: transformationContext.factory,
      program,
      rootDirs,
      transformationContext: transformationContext,
      tsInstance,
      emitHost,
      resolver,
      isTranspileOnly,
      isTsNode,
      indexChecker: new IndexChecker(tsInstance),
      excludeMatchers: config.exclude?.map((globPattern) => new Minimatch(globPattern, { matchBase: true })),
      outputFileNamesCache: new Map(),
      pathsPatterns:
        paths &&
        (tryParsePatterns
          ? configFile?.configFileSpecs?.pathPatterns || tryParsePatterns(paths)
          : tsInstance.getOwnKeys(paths)),
    };

    return (sourceFile: TS.SourceFile) => {
      const visitorContext: VisitorContext = {
        ...tsTransformPathsContext,
        sourceFile,
        isDeclarationFile: sourceFile.isDeclarationFile,
        originalSourceFile: (<typeof TS>tsInstance).getOriginalSourceFile(sourceFile),
        getVisitor() {
          return nodeVisitor.bind(this);
        },
        factory: createHarmonyFactory(tsInstance, transformationContext.factory),
      };

      return tsInstance.visitEachChild(sourceFile, visitorContext.getVisitor(), transformationContext);
    };
  };
}

// endregion
