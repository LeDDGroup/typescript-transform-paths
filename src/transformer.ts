import path from "node:path";
import ts, { CompilerOptions } from "typescript";
import { RunMode, TsNodeState, TsTransformPathsConfig, TsTransformPathsContext, VisitorContext } from "./types";
import { nodeVisitor } from "./visitor";
import { createHarmonyFactory } from "./harmony";
import { Minimatch } from "minimatch";
import { createSyntheticEmitHost, getTsNodeRegistrationProperties } from "./utils/ts-helpers";
import { TransformerExtras } from "ts-patch";

function getTsProperties(args: Parameters<typeof transformer>) {
  let fileNames: readonly string[] | undefined;
  let compilerOptions: CompilerOptions;
  let runMode: RunMode;
  let tsNodeState: TsNodeState | undefined;

  const { 0: program, 2: extras, 3: manualTransformOptions } = args;

  const tsInstance = extras?.ts ?? ts;

  if (program) compilerOptions = program.getCompilerOptions();
  const tsNodeProps = getTsNodeRegistrationProperties(tsInstance);

  /* Determine RunMode & Setup */
  // Note: ts-node passes a Program with the paths property stripped, so we do some comparison to determine if it's the caller
  const isTsNode =
    tsNodeProps && (!program || compilerOptions!.configFilePath === tsNodeProps.compilerOptions.configFilePath);

  // RunMode: Program
  if (program && !isTsNode) {
    runMode = RunMode.Program;
    compilerOptions = compilerOptions!;
  }
  // RunMode: Manual
  else if (manualTransformOptions) {
    runMode = RunMode.Manual;
    fileNames = manualTransformOptions.fileNames;
    compilerOptions = manualTransformOptions.compilerOptions!;
  }
  // RunMode: TsNode
  else if (isTsNode) {
    fileNames = tsNodeProps.fileNames;
    runMode = RunMode.TsNode;

    tsNodeState =
      !program ||
      (fileNames.length > 1 && program?.getRootFileNames().length === 1) ||
      (!compilerOptions!.paths && tsNodeProps!.compilerOptions.paths)
        ? TsNodeState.Stripped
        : TsNodeState.Full;

    compilerOptions =
      tsNodeState === TsNodeState.Full
        ? compilerOptions!
        : {
            ...program?.getCompilerOptions(),
            ...tsNodeProps!.compilerOptions,
          };
  } else {
    throw new Error(
      `Cannot transform without a Program, ts-node instance, or manual parameters supplied. ` +
        `Make sure you're using ts-patch or ts-node with transpileOnly.`,
    );
  }

  return { tsInstance, compilerOptions, fileNames, runMode, tsNodeState };
}

export default function transformer(
  program?: ts.Program,
  pluginConfig?: TsTransformPathsConfig,
  transformerExtras?: TransformerExtras,
  /** Supply if manually transforming with compiler API via 'transformNodes' / 'transformModule' */
  manualTransformOptions?: {
    compilerOptions?: ts.CompilerOptions;
    fileNames?: string[];
  },
) {
  return (transformationContext: ts.TransformationContext) => {
    // prettier-ignore
    const {
      tsInstance,
      compilerOptions,
      fileNames,
      runMode,
      tsNodeState
    } = getTsProperties([ program, pluginConfig, transformerExtras, manualTransformOptions ]);

    const rootDirs = compilerOptions.rootDirs?.filter(path.isAbsolute);
    const config: TsTransformPathsConfig = pluginConfig ?? {};
    const getCanonicalFileName = tsInstance.createGetCanonicalFileName(tsInstance.sys.useCaseSensitiveFileNames);

    /* Add supplements for various run modes */
    let emitHost = transformationContext.getEmitHost();
    if (!emitHost || tsNodeState === TsNodeState.Stripped) {
      if (!fileNames)
        throw new Error(
          `No EmitHost found and could not determine files to be processed. Please file an issue with a reproduction!`,
        );
      emitHost = createSyntheticEmitHost(compilerOptions, tsInstance, getCanonicalFileName, fileNames as string[]);
    }

    /* Create Visitor Context */
    const { configFile, paths } = compilerOptions;
    const { tryParsePatterns } = tsInstance;
    const [tsVersionMajor, tsVersionMinor] = tsInstance.versionMajorMinor.split(".").map((v) => +v);

    if (!tsVersionMajor || !tsVersionMinor) throw new Error("Expected version to be parsed");

    const tsTransformPathsContext: TsTransformPathsContext = {
      compilerOptions,
      config,
      elisionMap: new Map(),
      tsFactory: transformationContext.factory,
      program,
      rootDirs,
      transformationContext,
      tsInstance,
      tsVersionMajor,
      tsVersionMinor,
      emitHost,
      runMode,
      tsNodeState,
      excludeMatchers: config.exclude?.map((globPattern) => new Minimatch(globPattern, { matchBase: true })),
      outputFileNamesCache: new Map(),
      // Get paths patterns appropriate for TS compiler version
      pathsPatterns:
        paths &&
        (tryParsePatterns
          ? configFile?.configFileSpecs?.pathPatterns || tryParsePatterns(paths)
          : tsInstance.getOwnKeys(paths)),
    };

    return (sourceFile: ts.SourceFile) => {
      const visitorContext: VisitorContext = {
        ...tsTransformPathsContext,
        sourceFile,
        isDeclarationFile: sourceFile.isDeclarationFile,
        originalSourceFile: ts.getParseTreeNode(sourceFile, ts.isSourceFile) || sourceFile,
        getVisitor() {
          return nodeVisitor.bind(this);
        },
        factory: createHarmonyFactory(tsTransformPathsContext),
      };

      return tsInstance.visitEachChild(sourceFile, visitorContext.getVisitor(), transformationContext);
    };
  };
}
