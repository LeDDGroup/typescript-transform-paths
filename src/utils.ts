import path from "path";
import url from 'url';

/* ****************************************************************************************************************** *
 * TS Exports
 * ****************************************************************************************************************** */

import { normalizePath } from 'typescript';
export { removeFileExtension, isAnySupportedFileExtension, removeSuffix } from 'typescript';
export { normalizePath };

/* ****************************************************************************************************************** */
// region: General Utils
/* ****************************************************************************************************************** */

export const isURL = (s: string): boolean => !!s && (!!url.parse(s).host || !!url.parse(s).hostname);
export const cast = <T>(v: any): T => v;
export const isBaseDir = (baseDir: string, testDir: string): boolean => {
  const relative = path.relative(baseDir, testDir);
  return relative ? !relative.startsWith("..") && !path.isAbsolute(relative) : true;
};
export const maybeAddRelativeLocalPrefix = (p: string) => (p[0] === "." ? p : `./${p}`);

// endregion

/* ****************************************************************************************************************** */
// region: Path Utils
/* ****************************************************************************************************************** */

/**
 * Merge and normalize paths (preserves relative prefix – ie. `./my/path`)
 */
export function joinPaths(...paths: (string | undefined)[]): string {
  // path.join / normalizePath cannot be used here, because they remove relative prefix
  return normalizeSlashes((paths.filter(p => !!p) as string[]).join('/'))!;
}

/**
 * Normalize slashes in path
 */
export function normalizeSlashes<T extends string | undefined>(p: T, opt?: { removeLeadingSlash?: boolean, removeTrailingSlash?: boolean }): T {
  if (!p) return p;

  let res = p.replace(/\\+|\/+/g, '/');

  if (opt?.removeLeadingSlash && res[0] === '/') res = res.slice(1);
  if (opt?.removeTrailingSlash && res[res.length-1] === '/') res = res.slice(0, res.length - 1);

  return res as T;
}

/**
 * Get extension for path (supports declaration style – ie. `d.ts`)
 */
export function extName(p: string) {
  const res = path.extname(p);
  const baseName = path.basename(p, res);
  return baseName.slice(-2) === '.d' ? `.d${res}` : res;
}

export function baseName(p: string, stripExtension: boolean | string = false) {
  const ext = !stripExtension ? void 0 : (typeof stripExtension === 'string' ? stripExtension : extName(p));
  return path.basename(p, ext);
}

export function resolvePath(...p: (string | undefined)[]): string {
  return normalizePath(path.resolve(joinPaths(...p))) as string;
}

export function dirName(p: string) {
  return normalizePath(path.dirname(p));
}

export function relativePath(from: string, to: string) {
  return normalizePath(path.relative(from, to));
}

// endregion
