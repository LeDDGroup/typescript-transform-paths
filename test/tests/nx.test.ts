import { execSync } from "node:child_process";
import { rmSync } from "node:fs";
import path from "node:path";

import ts from "typescript";

import { nxTransformerPlugin } from "typescript-transform-paths";
import * as transformerModule from "../../dist/transformer";

import { projectsPaths } from "../config";
import { readTree } from "../utils";

describe(`NX Transformer`, () => {
  describe("Plugin", () => {
    let mockedTransformer: jest.SpyInstance;

    const program = { x: 1 };

    beforeAll(async () => {
      // @ts-expect-error TS(2345) FIXME: Argument of type '() => void' is not assignable to parameter of type '(transformationContext: TransformationContext) => (sourceFile: SourceFile) => SourceFile'.
      mockedTransformer = jest.spyOn(transformerModule, "default").mockReturnValue(() => {});
    });
    afterAll(() => {
      mockedTransformer.mockClear();
    });
    beforeEach(() => {
      mockedTransformer.mockReset();
    });

    test(`Before properly routes transform`, () => {
      const config = { a: 2 };

      // @ts-expect-error TS(2559) FIXME: Type '{ a: number; }' has no properties in common with type 'Omit<TsTransformPathsConfig, "transform">'.
      nxTransformerPlugin.before(config, program);

      expect(mockedTransformer).toHaveBeenCalledTimes(1);
      expect(mockedTransformer.mock.lastCall).toHaveLength(2);

      const [recProgram, recConfig] = mockedTransformer.mock.lastCall;
      expect(recProgram).toBe(program);
      expect(recConfig).toStrictEqual(config);
    });

    test(`After properly routes transform`, () => {
      const config = { a: 2, afterDeclarations: true };

      // @ts-expect-error TS(2345) FIXME: Argument of type '{ x: number; }' is not assignable to parameter of type 'Program'.
      nxTransformerPlugin.afterDeclarations(config, program);

      expect(mockedTransformer).toHaveBeenCalledTimes(1);
      expect(mockedTransformer.mock.lastCall).toHaveLength(2);

      const [recProgram, recConfig] = mockedTransformer.mock.lastCall;
      expect(recProgram).toBe(program);
      expect(recConfig).toStrictEqual({ ...config, afterDeclarations: true });
    });
  });

  describe(`(e2e) Works in NX project`, () => {
    const projectRoot = ts.normalizePath(path.join(projectsPaths, "nx"));

    // TODO - Investigate ways to do without emit
    test(`Transformer works for emitted declaration`, () => {
      execSync("yarn build", { cwd: projectRoot });

      try {
        expect(readTree(path.join(projectRoot, "dist"))).toMatchSnapshot();
      } finally {
        rmSync(`${projectRoot}/dist`, { recursive: true, force: true });
      }
    });
  });
});
