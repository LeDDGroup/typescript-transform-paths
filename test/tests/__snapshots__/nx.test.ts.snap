// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NX Transformer (e2e) Works in NX project Transformer works for emitted declaration 1`] = `
{
  "library1/index.js": "/******/ (() => { // webpackBootstrap
/******/ 	"use strict";
/******/ 	var __webpack_modules__ = ([
/* 0 */,
/* 1 */
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, "__esModule", ({ value: true }));
exports.name = void 0;
exports.name = "library2";


/***/ })
/******/ 	]);
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it uses a non-standard name for the exports (exports).
(() => {
var exports = __webpack_exports__;

Object.defineProperty(exports, "__esModule", ({ value: true }));
const library2_1 = __webpack_require__(1);
exports["default"] = library2_1.name;
console.log(library2_1.name);

})();

var __webpack_export_target__ = exports;
for(var i in __webpack_exports__) __webpack_export_target__[i] = __webpack_exports__[i];
if(__webpack_exports__.__esModule) Object.defineProperty(__webpack_export_target__, "__esModule", { value: true });
/******/ })()
;",
  "library1/packages/library1/src/index.d.ts": "import { name as library2Name } from "../../library2/src";
export default library2Name;
",
  "library1/packages/library2/src/index.d.ts": "export declare const name = "library2";
",
}
`;
